stdlib/Logging/docs/src/index.md

```diff
@@ -0,0 +1,202 @@
+```
+
+will produce no output by default.  Furthermore, it's very cheap to leave debug
+statements like this in the source code because the system avoids evaluating
+the message if it would later be ignored.  In this case `sum(rand(100))` and
+the associated string processing will never be executed unless debug logging is
+enabled.
+
+Second, the logging tools allow you to attach arbitrary data to each event as a
+set of key value pairs. This allows you to capture local variables and other
+program state for later analysis. For example, to attach the local array
+variable `A` and the sum of a vector `v` as the key `s` you can use
+
+```jldoctest
+A = ones(Int, 4, 4)
+v = ones(100)
+@info "Some variables"  A  s=sum(v)
+
+# Output
+┌ Info: Some variables
+│   A =
+│    4×4 Array{Int64,2}:
+│     1  1  1  1
+│     1  1  1  1
+│     1  1  1  1
+│     1  1  1  1
+└   s = 100.0
+```
+
+All of the logging macros `@debug`, `@info`, `@warn` and `@error` share common
+features which are described in detail in the documentation for the more
+general macro [`@logmsg`](@ref).
+
+## Log event structure
+
+Each event generates several pieces of data, some provided by the user and some
+automatically extracted. Let's examine the user-defined data first:
+
+* The *log level* is a broad category for the message which is used for early
+  filtering. There are several standard levels of type [`LogLevel`](@ref);
+  user defined levels are also possible.
+  - Use `Debug` for verbose information which could be useful when debugging an
+    application or module. These events are disabled by default.
+  - Use `Info` to inform the user about the normal operation of the program.
+  - Use `Warn` when a potential problem is detected.
+  - Use `Error` to report errors where the code has enough context to recover
+    and continue.  (When the code doesn't have enough context, an exception or
+    early return is more appropriate.)
+* The *message* is a human readable string describing the event in markdown
+  format.
+* Optional *key value pairs* allow arbitrary data to be attached to each event.
+  Some keys have conventional meaning which can affect the way an event is
+  interpreted (see [`@logmsg`](@ref)).
+
+The system also generates some standard information for each event:
+
+* The `module` in which the logging macro was expanded.
+* The `file` and `line` where the logging macro occurs in the source code.
+* A message `id` which is unique for each logging macro invocation. This is
+  very useful as a key for caching information or actions associated with an
+  event. For instance, it can be used to limit the number of times a message
+  is presented to the user.
+* A `group` for the event which is set to the base name of the file by default,
+  without extension.  This can be used to group messages into categories more
+  finely than the log level (for example, all deprecation warnings have group
+  `:depwarn`), or into logical groupings across or within modules.
```

Comment:
Is there any way to attach additional information to log events?

For example, in IJulia we might want to attach additional information about the current execution cell, so that the log message is printed in the right output cell.   (If necessary, we can extract this from the `file` metadata, since the input cell name is used as the filename, but I'm curious if there is a way to attach additional runtime information.  For example, if you define a custom logger, could you implement a method to generate additional key-value pairs based on some dynamic global state at the time the log message is emitted?)

Reply:
Absolutely — dynamic context is something which can be added by installing a custom logger as a filter (add a key–value pair and pass to the next logger up the stack).  This could include information such as:
* Global or task-local variables as you're suggesting.
* The time of the event
* A backtrace

I think this is going to be a common enough pattern that we could support it in the library. For now however it's pretty bare bones and provides just enough structure for this kind of stuff to be done outside Base/stdlib.

For IJulia I expect you'll want to install a custom logger anyway because formatting of log messages can be a lot nicer in the browser compared to the console.